Levelgenerator

AndreasBonny opened the pull request:
Level generation implemented for a basic generation with some normal rooms and some special rooms like a boss room or a vendor. 
The number of rooms is static at the moment but it shouldn't pose much of a problem to do it dynamically.

Commit:
prototype level generation

Commit:
Better level generation + Comments

Commit:
special rooms in level generation

Commit:
special rooms can't be next to another special room (exception is the secret room)

Commit:
special rooms can't be next to another special room (exception is the secret room) #2

Osmil reviewed (commented):
Wouldn't passing the normal_room count in the constructor be better and make the implementation more flexible?

Osmil commented on core/src/com/moreoptions/prototype/level/Map.java:
```diff
@@ -0,0 +1,286 @@
+    private boolean isStartingRoom = true;
+
+
+    private int roomCounter = 0;
```
Maybe pass a max_normal_room in the constructor.

Osmil commented on core/src/com/moreoptions/prototype/level/Map.java:
```diff
@@ -0,0 +1,286 @@
+     */
+    private void fillMap() {
+        // TODO: better closure for while loop (not maxRooms + X)
+        while (roomCounter <= maxRooms + 4) {
```
4 is a constant here, because we always generate 1 treasure, 1 boss, 1 special and 1 vendor.

Osmil commented on core/src/com/moreoptions/prototype/level/Map.java:
```diff
@@ -0,0 +1,286 @@
+    public Map() {
+        this.map = new int[width + 1][height + 1];
+        createMap();
+        fillMap();
```
runtime is somewhere between 15+4 or infinity

Commit:
another map implementation which should work faster than the old one.
implementation still has some problems but the algorithm generally works

Commit:
bounds are accounted for (can create 50+ rooms on a 10x10 field without breaking)

Commit:
interface for the methods

new implementation is not done yet

Commit:
map generation finished?
can create every room within the bounds of the map without throwing an error, room counter is changeable

Commit:
some tests for the minimap implementation

refactoring and javadoc for the minimap

Commit:
some tests for the minimap implementation

refactoring and javadoc for the minimap

(without DesktopLauncher)

Commit:
Merge remote-tracking branch 'origin/levelgenerator' into levelgenerator

Osmil committed:
Refactored a bit, implementation is still very sloppy. Will have to get back to this
